Tag generation fails for Scala Array

Scala version: 3.4.2
Izumi reflect version: 2.3.8 (transitively from zio-mock)

Writing a ZIO Mock using the Scala Array type makes the compiler crash with:

<details>

<summary>Compiler output</summary>

```
java.lang.AssertionError: assertion failed: class Array
	at scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:8)
	at dotty.tools.backend.jvm.BCodeHelpers.primitiveOrClassToBType$1(BCodeHelpers.scala:717)
	at dotty.tools.backend.jvm.BCodeHelpers.dotty$tools$backend$jvm$BCodeHelpers$$typeToTypeKind(BCodeHelpers.scala:737)
	at dotty.tools.backend.jvm.BCodeHelpers$BCInnerClassGen.toTypeKind(BCodeHelpers.scala:202)
	at dotty.tools.backend.jvm.BCodeHelpers$BCInnerClassGen.toTypeKind$(BCodeHelpers.scala:130)
	at dotty.tools.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.toTypeKind(BCodeSkelBuilder.scala:134)
	at dotty.tools.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genConstant(BCodeBodyBuilder.scala:582)
	at dotty.tools.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genLoadTo(BCodeBodyBuilder.scala:451)
	at dotty.tools.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genLoad(BCodeBodyBuilder.scala:304)
	at dotty.tools.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.loop$1(BCodeBodyBuilder.scala:1210)
	at dotty.tools.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genLoadArguments(BCodeBodyBuilder.scala:1217)
	at dotty.tools.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genApply(BCodeBodyBuilder.scala:836)
	at dotty.tools.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genLoadTo(BCodeBodyBuilder.scala:386)
	at dotty.tools.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genLoadTo(BCodeBodyBuilder.scala:466)
	at dotty.tools.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genLoad(BCodeBodyBuilder.scala:304)
	at dotty.tools.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genLoadTo(BCodeBodyBuilder.scala:321)
	at dotty.tools.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genLoad(BCodeBodyBuilder.scala:304)
	at dotty.tools.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genStat(BCodeBodyBuilder.scala:116)
	at dotty.tools.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genBlockTo$$anonfun$1(BCodeBodyBuilder.scala:1094)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at dotty.tools.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genBlockTo(BCodeBodyBuilder.scala:1094)
	at dotty.tools.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genLoadTo(BCodeBodyBuilder.scala:458)
	at dotty.tools.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genLoad(BCodeBodyBuilder.scala:304)
	at dotty.tools.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.loop$1(BCodeBodyBuilder.scala:1210)
	at dotty.tools.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genLoadArguments(BCodeBodyBuilder.scala:1217)
	at dotty.tools.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genApply(BCodeBodyBuilder.scala:836)
	at dotty.tools.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genLoadTo(BCodeBodyBuilder.scala:386)
	at dotty.tools.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genLoad(BCodeBodyBuilder.scala:304)
	at dotty.tools.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genStat(BCodeBodyBuilder.scala:116)
	at dotty.tools.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genBlockTo$$anonfun$1(BCodeBodyBuilder.scala:1094)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at dotty.tools.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genBlockTo(BCodeBodyBuilder.scala:1094)
	at dotty.tools.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genLoadTo(BCodeBodyBuilder.scala:458)
	at dotty.tools.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.emitNormalMethodBody$1(BCodeSkelBuilder.scala:893)
	at dotty.tools.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.genDefDef(BCodeSkelBuilder.scala:916)
	at dotty.tools.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.gen(BCodeSkelBuilder.scala:693)
	at dotty.tools.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.gen$$anonfun$1(BCodeSkelBuilder.scala:699)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at dotty.tools.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.gen(BCodeSkelBuilder.scala:699)
	at dotty.tools.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.genPlainClass(BCodeSkelBuilder.scala:293)
	at dotty.tools.backend.jvm.CodeGen.genClass(CodeGen.scala:154)
	at dotty.tools.backend.jvm.CodeGen.genClassDef$1(CodeGen.scala:62)
	at dotty.tools.backend.jvm.CodeGen.genClassDefs$1(CodeGen.scala:117)
	at dotty.tools.backend.jvm.CodeGen.genClassDefs$1$$anonfun$1(CodeGen.scala:115)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at dotty.tools.backend.jvm.CodeGen.genClassDefs$1(CodeGen.scala:115)
	at dotty.tools.backend.jvm.CodeGen.genUnit(CodeGen.scala:120)
	at dotty.tools.backend.jvm.GenBCode.run(GenBCode.scala:83)
	at dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:354)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:360)
	at dotty.tools.backend.jvm.GenBCode.runOn(GenBCode.scala:91)
	at dotty.tools.dotc.Run.runPhases$1$$anonfun$1(Run.scala:315)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1323)
	at dotty.tools.dotc.Run.runPhases$1(Run.scala:337)
	at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:350)
	at dotty.tools.dotc.Run.compileUnits$$anonfun$adapted$1(Run.scala:360)
	at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:69)
	at dotty.tools.dotc.Run.compileUnits(Run.scala:360)
	at dotty.tools.dotc.Run.compileSources(Run.scala:261)
	at dotty.tools.dotc.Run.compile(Run.scala:246)
	at dotty.tools.dotc.Driver.doCompile(Driver.scala:37)
	at dotty.tools.xsbt.CompilerBridgeDriver.run(CompilerBridgeDriver.java:141)
	at dotty.tools.xsbt.CompilerBridge.run(CompilerBridge.java:22)
	at sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:91)
	at sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$7(MixedAnalyzingCompiler.scala:193)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at sbt.internal.inc.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:248)
	at sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4(MixedAnalyzingCompiler.scala:183)
	at sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4$adapted(MixedAnalyzingCompiler.scala:163)
	at sbt.internal.inc.JarUtils$.withPreviousJar(JarUtils.scala:239)
	at sbt.internal.inc.MixedAnalyzingCompiler.compileScala$1(MixedAnalyzingCompiler.scala:163)
	at sbt.internal.inc.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:211)
	at sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1(IncrementalCompilerImpl.scala:534)
	at sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1$adapted(IncrementalCompilerImpl.scala:534)
	at sbt.internal.inc.Incremental$.$anonfun$apply$5(Incremental.scala:180)
	at sbt.internal.inc.Incremental$.$anonfun$apply$5$adapted(Incremental.scala:178)
	at sbt.internal.inc.Incremental$$anon$2.run(Incremental.scala:464)
	at sbt.internal.inc.IncrementalCommon$CycleState.next(IncrementalCommon.scala:116)
	at sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:56)
	at sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:52)
	at sbt.internal.inc.IncrementalCommon.cycle(IncrementalCommon.scala:263)
	at sbt.internal.inc.Incremental$.$anonfun$incrementalCompile$8(Incremental.scala:419)
	at sbt.internal.inc.Incremental$.withClassfileManager(Incremental.scala:506)
	at sbt.internal.inc.Incremental$.incrementalCompile(Incremental.scala:406)
	at sbt.internal.inc.Incremental$.apply(Incremental.scala:172)
	at sbt.internal.inc.IncrementalCompilerImpl.compileInternal(IncrementalCompilerImpl.scala:534)
	at sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileIncrementally$1(IncrementalCompilerImpl.scala:488)
	at sbt.internal.inc.IncrementalCompilerImpl.handleCompilationError(IncrementalCompilerImpl.scala:332)
	at sbt.internal.inc.IncrementalCompilerImpl.compileIncrementally(IncrementalCompilerImpl.scala:425)
	at sbt.internal.inc.IncrementalCompilerImpl.compile(IncrementalCompilerImpl.scala:137)
	at sbt.Defaults$.compileIncrementalTaskImpl(Defaults.scala:2371)
	at sbt.Defaults$.$anonfun$compileIncrementalTask$2(Defaults.scala:2321)

	at sbt.internal.server.BspCompileTask$.$anonfun$compute$1(BspCompileTask.scala:31)
```
</details>

The error comes from the makeTag Macro.
Here is a demo: https://scastie.scala-lang.org/ugTtSPC9STqinqMGS1nieQ

--- a/izumi-reflect/izumi-reflect/src/main/scala-2/izumi/reflect/TagMacro.scala
+++ b/izumi-reflect/izumi-reflect/src/main/scala-2/izumi/reflect/TagMacro.scala
@@ -397,18 +397,24 @@ class TagMacro(val c: blackbox.Context) {
   @inline
   private[this] final def closestClass(properTypeStrongCtor: Type): c.Expr[Class[_]] = {
     // unfortunately .erasure returns trash for intersection types
-    val tpeLub = properTypeStrongCtor match {
+    val tpeLub = ReflectionUtil.norm(c.universe: c.universe.type, logger)(properTypeStrongCtor.dealias) match {
       case r: RefinedTypeApi => lub(r.parents)
-      case _ => properTypeStrongCtor
+      case o => o
     }
     val tpeErased = tpeLub.erasure
     // and for Scala varargs (Scala by names and Java varargs are fine)
     val tpeFixed = if (tpeErased.typeSymbol eq definitions.RepeatedParamClass) {
       typeOf[scala.Seq[Any]].dealias
+    } else if (tpeErased.typeSymbol eq definitions.ArrayClass) {
+      // workaround for a crash that happens when .erasure misfires on Array in case `Tag[Array[List[X]]]`
+      // and produces a tree `classOf[Array[List]]` which fails to compile.
+      // Array is the only type that needs to be parameterized after erasure and stripping its parameters via .erasure
+      // actually breaks it, so we skip this step for Arrays.
+      tpeLub.dealias
     } else {
       tpeErased
     }
-    c.Expr[Class[_]](q"_root_.scala.Predef.classOf[$tpeFixed]")
+    c.Expr[Class[_]](q"${Literal(Constant(tpeFixed))}.asInstanceOf[_root_.java.lang.Class[_]]")
   }
 
   @inline

--- a/izumi-reflect/izumi-reflect/src/main/scala-3/izumi/reflect/TagMacro.scala
+++ b/izumi-reflect/izumi-reflect/src/main/scala-3/izumi/reflect/TagMacro.scala
@@ -34,7 +34,7 @@ final class TagMacro(using override val qctx: Quotes) extends InspectorBase {
         // There's probably a missing case where a higher-kinded type is not a type lambda, with splicing inbetween causing fixup by the compiler
         //   val ltt = Inspect.inspectAny[A]
         val ltt = '{ Inspect.inspect[a] }
-        val cls = closestClassOfExpr(typeRepr)
+        val cls = closestClassOfTypeRepr(typeRepr)
         '{ Tag[a]($cls, $ltt) }.asInstanceOf[Expr[Tag[A]]]
     }
   }
@@ -159,7 +159,7 @@ final class TagMacro(using override val qctx: Quotes) extends InspectorBase {
       case andType: AndType =>
         val tpes = flattenAnd(andType)
         val ltts: Expr[List[LightTypeTag]] = Expr.ofList(tpes.map(summonLTTAndFastTrackIfNotTypeParam))
-        val cls = Literal(ClassOfConstant(lubClassOf(tpes))).asExprOf[Class[?]]
+        val cls = Literal(ClassOfConstant(lubClassOf(typeReprDealiased, tpes))).asExprOf[Class[?]]
         val dummyAnyStructLtt = {
           // FIXME add constructor for intersections without the unused on Scala 3 struct type
           Inspect.inspectAny[Any]
@@ -169,12 +169,12 @@ final class TagMacro(using override val qctx: Quotes) extends InspectorBase {
       case orType: OrType =>
         val tpes = flattenOr(orType)
         val ltts: Expr[List[LightTypeTag]] = Expr.ofList(tpes.map(summonLTTAndFastTrackIfNotTypeParam))
-        val cls = Literal(ClassOfConstant(lubClassOf(tpes))).asExprOf[Class[?]]
+        val cls = Literal(ClassOfConstant(lubClassOf(typeReprDealiased, tpes))).asExprOf[Class[?]]
         '{ Tag.unionTag[T](${ cls }, ${ ltts }) }
 
       case refinement: Refinement =>
         val (members, parent) = flattenRefinements(refinement)
-        val cls = closestClassOfExpr(parent)
+        val cls = closestClassOfTypeRepr(parent)
         val parentLtt = summonLTTAndFastTrackIfNotTypeParam(parent)
 
         val (allTypeMembers, termMembers) = members.partitionMap {
@@ -229,17 +229,23 @@ final class TagMacro(using override val qctx: Quotes) extends InspectorBase {
     }
   }
 
-  private def closestClassOfExpr(typeRepr: TypeRepr): Expr[Class[?]] = {
-    Literal(ClassOfConstant(lubClassOf(intersectionUnionRefinementClassPartsOf(typeRepr)))).asExprOf[Class[?]]
+  private def closestClassOfTypeRepr(typeRepr: TypeRepr): Expr[Class[?]] = {
+    Literal(ClassOfConstant(lubClassOf(typeRepr, intersectionUnionRefinementClassPartsOf(typeRepr)))).asExprOf[Class[?]]
   }
 
-  private def lubClassOf(tpes: List[TypeRepr]): TypeRepr = {
+  private def lubClassOf(specificTpe: TypeRepr, tpes: List[TypeRepr]): TypeRepr = {
     tpes.map(_.baseClasses) match {
       case h :: t =>
         val bases = h.to(mutable.LinkedHashSet)
-        t.foreach(b => bases.filterInPlace(b.to(mutable.HashSet).contains))
+        t.foreach { b =>
+          val bBases = b.to(mutable.HashSet)
+          bases.filterInPlace(bBases)
+        }
         // rely on the fact that .baseClasses returns classes in order from most specific to least, therefore most specific class should be first.
-        bases.headOption.getOrElse(defn.AnyClass).typeRef
+        val baseClass = bases.headOption.getOrElse(defn.AnyClass)
+        // try to recover type parameters of the specific type e.g. to support Arrays
+        // see https://github.com/zio/izumi-reflect/issues/474 & https://github.com/scala/scala3/issues/21916
+        specificTpe.baseType(baseClass)
       // FIXME: below doesn't work, need to treat AnyVals specially
 //        bases.find(!_.typeRef.baseClasses.contains(defn.AnyValClass)).getOrElse(defn.AnyClass).typeRef
       case Nil =>

--- a/izumi-reflect/izumi-reflect/src/test/scala/izumi/reflect/test/SharedTagTest.scala
+++ b/izumi-reflect/izumi-reflect/src/test/scala/izumi/reflect/test/SharedTagTest.scala
@@ -30,6 +30,7 @@ trait ZY extends Assertions {
   type U = T
   type V = List[T]
   type A = List[Option[Int]]
+  type O <: List[T]
   val x: String = "5"
   object y
   trait Y
@@ -649,15 +650,52 @@ abstract class SharedTagTest extends AnyWordSpec with XY[String] with TagAsserti
 
       assert(Tag[String].closestClass ne classOf[AnyVal])
       assert(!Tag[String with Int].hasPreciseClass)
+      assert(Tag[String with Int].closestClass eq classOf[Any])
 
-      assert(Tag[List[Int]].closestClass eq classOf[List[_]])
       assert(Tag[List[Int]].hasPreciseClass)
-      assert(Tag[H1].hasPreciseClass)
+      assert(Tag[List[Int]].closestClass eq classOf[List[_]])
 
-      assert(Tag[ZY#T].closestClass eq classOf[Any])
+      assert(Tag[H1].hasPreciseClass)
+      assert(Tag[H1].closestClass eq classOf[H1])
+
       assert(!Tag[ZY#T].hasPreciseClass)
+      assert(Tag[ZY#T].closestClass eq classOf[Any])
 
       assert(Tag[ZY#Y].hasPreciseClass)
+      assert(Tag[ZY#Y].closestClass eq classOf[ZY#Y])
+
+      assert(!Tag[ZY#O].hasPreciseClass)
+      assert(Tag[ZY#O].closestClass eq classOf[List[ZY#T]])
+
+      assert(Tag[Array[Byte]].closestClass eq classOf[Array[Byte]])
+      assert(Tag[Array[Int]].closestClass eq classOf[Array[Int]])
+      assert(Tag[Array[AnyRef]].closestClass eq classOf[Array[AnyRef]])
+      assert(Tag[Array[Boolean]].closestClass eq classOf[Array[Boolean]])
+      assert(Tag[Array[Char]].closestClass eq classOf[Array[Char]])
+      assert(Tag[Array[String]].closestClass eq classOf[Array[String]])
+      assert(Tag[Array[List[Any]]].closestClass eq classOf[Array[List[Any]]])
+
+      object test1 {
+        sealed trait TX0
+        class TX1 extends TX0
+        class TX2 extends TX0
+        class TX3 extends TX0
+
+        type OX[T] = TX1 with TX2 with T
+      }
+
+      assert(!Tag[test1.OX[ZY#Y]].hasPreciseClass)
+      assert(!Tag[test1.OX[test1.TX3]].hasPreciseClass)
+      assert(!Tag[test1.OX[test1.TX0]].hasPreciseClass)
+      assert(Tag[test1.OX[ZY#Y]].closestClass eq classOf[Any])
+      assert(Tag[test1.OX[test1.TX3]].closestClass eq classOf[test1.TX0])
+      assert(Tag[test1.OX[test1.TX0]].closestClass eq classOf[test1.TX0])
+
+      object test2 {
+        type ArrayT <: Array[Byte]
+      }
+      assert(!Tag[test2.ArrayT].hasPreciseClass)
+      assert(Tag[test2.ArrayT].closestClass eq classOf[Array[Byte]])
     }
 
     "Work with term type prefixes" in {
@@ -1049,6 +1087,10 @@ abstract class SharedTagTest extends AnyWordSpec with XY[String] with TagAsserti
       assertDebugSame(Tag[(Object {}) @IdAnnotation("x") with Option[(String with Object) {}]].tag, LTT[Option[String]])
       assertDebugSame(Tag[(Any {}) @IdAnnotation("x") with Option[(String with Object) {}]].tag, LTT[Option[String]])
       assertDebugSame(Tag[(AnyRef {}) @IdAnnotation("x") with Option[(String with Object) {}]].tag, LTT[Option[String]])
+    }
+
+    "regression test for https://github.com/zio/izumi-reflect/issues/474" in {
+      assertCompiles("Tag[Array[Byte]]")
     }
 
   }

